src/types/basic.rs

```diff
@@ -0,0 +1,160 @@
+#![allow(missing_docs)]
+
+use super::{AbstractSignature, CustomType, TypeTag};
+
+enum EqTypeImpl {
+    USize,
+}
+enum ClassicTypeImpl {
+    E(EqTypeImpl),
+    Graph(Box<AbstractSignature>),
+}
+enum AnyTypeImpl {
+    C(ClassicTypeImpl),
+}
+
+pub struct Eq(EqTypeImpl);
+pub struct Classic(ClassicTypeImpl);
+pub struct Any(AnyTypeImpl);
+
+mod sealed {
+    use super::{Any, Classic, Eq};
+    pub trait Sealed {}
+    impl Sealed for Any {}
+    impl Sealed for Classic {}
+    impl Sealed for Eq {}
+}
+pub trait TypeClass: sealed::Sealed {
+    const TAG: TypeTag;
+}
+
+impl TypeClass for Eq {
+    const TAG: TypeTag = TypeTag::Hashable;
+}
+
+impl TypeClass for Classic {
+    const TAG: TypeTag = TypeTag::Classic;
+}
+impl TypeClass for Any {
+    const TAG: TypeTag = TypeTag::Simple;
+}
+pub struct TaggedWrapper<T>(TypeTag, T);
+
+impl<T> TaggedWrapper<T> {
+    pub fn tag(&self) -> TypeTag {
+        self.0
+    }
+
+    pub fn inner(&self) -> &T {
+        &self.1
+    }
+}
+pub enum Type<T> {
+    Prim(T),
+    Extension(CustomType),
+    Alias(TaggedWrapper<String>),
+    Array(Box<Type<T>>, usize),
+    Tuple(Vec<Type<T>>),
+    Sum(Vec<Type<T>>),
+}
+
+impl<T: TypeClass> Type<T> {
+    pub const fn tag(&self) -> TypeTag {
+        T::TAG
+    }
+
+    pub fn new_tuple(types: impl IntoIterator<Item = Type<T>>) -> Self {
+        Self::Tuple(types.into_iter().collect())
+    }
+
+    pub fn new_opaque(opaque: CustomType) -> Self {
+        assert!(opaque.tag() == T::TAG, "tag ");
```

Comment:
`T::TAG.contains(opaque.tag())`, I think?

Reply:
(And similarly we'd want a new_alias that makes the same check on the provided `TaggedWrapper<String>`)